x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/tokenizers/BpeTokenizer.java

```diff
@@ -0,0 +1,418 @@
+                            minSeq = sequence;
+                            minRank = rank;
+                        }
+                    }
+                    if (minSeq == null) {
+                        break;
+                    }
+                    List<CharSequence> mergedBpeTokens = new ArrayList<>(bpeTokens.size() - 1);
+                    for (int i = 0; i < minSeq.firstPos; i++) {
+                        mergedBpeTokens.add(bpeTokens.get(i));
+                    }
+                    mergedBpeTokens.add(minSeq);
+                    for (int i = minSeq.secondPos + 1; i < bpeTokens.size(); i++) {
+                        mergedBpeTokens.add(bpeTokens.get(i));
+                    }
+                    bpeTokens = mergedBpeTokens;
+                }
+                boolean subWordToken = false;
+                for (CharSequence charSequence : bpeTokens) {
+                    // call toString here to do a copy of the char[] array.
+                    // It is dangerous to have each token continue to share the same underlying array
+                    Integer tokenId = vocabulary.get(charSequence);
+                    // If this is the start of a new set of sub-word tokens AND it starts with a space, adjust the offsets to not include
+                    // the space. But, don't consider our potentially added space as it is not part of the original string
+                    int startOffsetAdj = subWordToken == false && charSequence.charAt(0) == ENCODED_SPACE_CHAR && addedSpace == false
+                        ? 1
+                        : 0;
+                    BpeToken toAdd = tokenId == null
+                        ? new BpeToken(
+                            unknownToken,
+                            subWordToken,
+                            tokenizedUnknown,
+                            offsetStart + offsetOffset + startOffsetAdj,
+                            offsetEnd + offsetOffset
+                        )
+                        : new BpeToken(
+                            charSequence.toString(),
+                            subWordToken,
+                            tokenId,
+                            offsetStart + offsetOffset + startOffsetAdj,
+                            offsetEnd + offsetOffset
+                        );
+                    tokens.add(toAdd);
+                    subWordToken = true;
+                }
+            }
+        }
+        filled = true;
+    }
+
+    private static List<CharSequencePair> pairs(List<CharSequence> tokens) {
+        List<CharSequencePair> pairs = new ArrayList<>(tokens.size() - 1);
+        for (int i = 0; i < tokens.size() - 1; i++) {
+            pairs.add(new CharSequencePair(MultiCharSequence.from(tokens.get(i), tokens.get(i + 1)), i, i + 1));
+        }
+        return pairs;
+    }
+
+    private void fillBuffer(Reader input) throws IOException {
+        int len;
+        str.setLength(0);
+        while ((len = input.read(buffer)) > 0) {
```

Comment:
Its typical for Lucene to have the buffer allocated once and reused throughout the life of the analyzer. Other wise, on ever string we read in, we reallocate a new `8192` length buffer. 

But, keep in mind, once we move away from Regex into a streaming approach (which I have ready as a fast follow), this will be moot.

Reply:
8192 is a large buffer to keep around especially as it just serves as a temporary holder to read the input into before appending it to the string builder. The string builder is the thing that should be allocated only once, a much smaller buffer (512 or 1024) could be used as the temporary.